test/adapters/browser/run-wasi.mjs

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env node
+
+import fs from 'fs/promises';
+import path from 'path';
+import { chromium } from "playwright"
+import { parseArgs } from "../shared/parseArgs.mjs"
+import { walkFs } from "../shared/walkFs.mjs"
+
+async function derivePreopens(dirs) {
+  const preopens = [];
+  for (let dir of dirs) {
+    const contents = await walkFs(dir, (name, entry, out) => {
+      if (entry.kind === "file") {
+        // Convert buffer to array to make it serializable.
+        entry.buffer = Array.from(entry.buffer);
+      }
+      return { ...out, [name]: entry };
+    }, {});
+    preopens.push({ dir, contents });
+  }
+  return preopens;
+}
+
+/**
+ * Configure routes for the browser harness.
+ *
+ * @param {import('playwright').BrowserContext} context 
+ * @param {string} harnessURL 
+ */
+async function configureRoutes(context, harnessURL) {
+
+  // Serve the main test page.
+  context.route(`${harnessURL}/run-test.html`, async route => {
+    const dirname = new URL(".", import.meta.url).pathname;
+    const body = await fs.readFile(path.join(dirname, "run-test.html"), "utf8");
+    route.fulfill({
+      status: 200,
+      contentType: 'text/html',
+      // Browsers reduce the precision of performance.now() if the page is not
+      // isolated. To keep the precision for `clock_get_time` we need to set the
+      // following headers.
+      // See: https://developer.mozilla.org/en-US/docs/Web/API/Performance/now#security_requirements
+      headers: {
+        "Cross-Origin-Opener-Policy": "same-origin",
+        "Cross-Origin-Embedder-Policy": "require-corp",
+      },
+      body,
+    });
+  })
+
+  // Serve wasi-testsuite files.
+  // e.g. http://browser-wasi-shim.localhost/home/me/browser_wasi_shim/test/wasi-testsuite/xxx
+  let projectDir = path.join(new URL("../../wasi-testsuite", import.meta.url).pathname);
+  projectDir = path.resolve(projectDir);
+  context.route(`${harnessURL}${projectDir}/**/*`, async route => {
```

Comment:
Are these only accessible to the headless browser or does playwright launch an actual webserver that is potentially exposed to other users on the same system or even the entire network?

Reply:
Based on a bit of looking around in the source code, it looks like it is only accessible to the headless browser.

Reply:
As far as I checked, the Playwright's network looks like using Chrome DevTools Protocol to intercept network requests triggered in browser page: https://github.com/microsoft/playwright/blob/b33a04bc880edd3473e07991719475b09251a192/packages/playwright-core/src/server/chromium/crNetworkManager.ts#L121-L125

So it does not launch web server, I think.